### PR TITLE
[ADD][10.0]BR001209-T24384: Business Requirement Gap Analysis Task

### PR DESCRIPTION
**Pull request by JoJoJoJoJoJoJo**
Add gap analysis task in business view and a wizard to create the gap analysis task

### CODE REVIEW

File: business_requirement_gap_analysis_task/README.rst [hunk 1]
```diff
@@ -0,0 +1,63 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.png
+   :target: http://www.gnu.org/licenses/agpl-3.0-standalone.html
+   :alt: License: AGPL-3
+
+======================================
+Business Requirement Gap Analysis Task
+======================================
+
+This module extends the functionality of Business Requirement to support gap
+analysis and to allow you to find gap analysis task easily and create it
```
Comment (elicoidal) [inline business_requirement_gap_analysis_task/README.rst]: to find gap analysis task easily and create it automatically.
=> to create gap analysis task and easily find it

File: business_requirement_gap_analysis_task/README.rst [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+automatically.
+
+
+Bug Tracker
+===========
+
+Bugs are tracked on `GitHub Issues
+<https://github.com/Elico-Corp/{project_repo}/issues>`_. In case of trouble, please
```
Comment (elicoidal) [inline business_requirement_gap_analysis_task/README.rst]: please adapt

File: business_requirement_gap_analysis_task/README.rst [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+   :alt: License: AGPL-3
+
+======================================
+Business Requirement Gap Analysis Task
+======================================
+
+This module extends the functionality of Business Requirement to support gap
+analysis and to allow you to find gap analysis task easily and create it
+automatically.
+
```
Comment (elicoidal) [inline business_requirement_gap_analysis_task/README.rst]: I miss the runbot information

File: business_requirement_gap_analysis_task/README.rst [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+check there if your issue has already been reported. If you spotted it first,
+help us smash it by providing detailed and welcomed feedback.
+
+Credits
+=======
+
+Images
+------
+
+* Odoo Community Association: `Icon <https://github.com/OCA/maintainer-tools/blob/master/template/module/static/description/icon.svg>`_.
```
Comment (elicoidal) [inline business_requirement_gap_analysis_task/README.rst]: use same icon as the other modules for business_requirements

File: business_requirement_gap_analysis_task/README.rst [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+.. only if the OCA logo is used, else remove
+
+Contributors
+------------
+
+* Firstname Lastname <email.address@example.org>
```
Comment (elicoidal) [inline business_requirement_gap_analysis_task/README.rst]: Please adapt

File: business_requirement_gap_analysis_task/README.rst [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+* Second Person <second.person@example.org>
+
+Maintainer
```
Comment (elicoidal) [inline business_requirement_gap_analysis_task/README.rst]: Use standard OCA template information

File: business_requirement_gap_analysis_task/__manifest__.py [hunk 7]
```diff
@@ -0,0 +1,31 @@
+# -*- coding: utf-8 -*-
+# © 2017 Elico Corp (https://www.elico-corp.com).
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+{
+    'name': "Business Requirement Gap Analysis Task",
+
```
Comment (elicoidal) [inline business_requirement_gap_analysis_task/__manifest__.py]: remove empty lines

File: business_requirement_gap_analysis_task/__manifest__.py [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+    'summary': """Add Gap Analysis Task to your Business Requirement""",
+
+    'description': """
```
Comment (elicoidal) [inline business_requirement_gap_analysis_task/__manifest__.py]: remove: only needed before v8

File: business_requirement_gap_analysis_task/__manifest__.py [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+        Generate this GA task automatically and link it to the BR so that it's \
+        easy to find
+    """,
+
+    'author': "Elico Corp, Odoo Community Association (OCA)",
+    'website': "https://www.elico-corp.com",
```
Comment (elicoidal) [inline business_requirement_gap_analysis_task/__manifest__.py]: use the BR repo URL: https://github.com/OCA/business-requirement

File: business_requirement_gap_analysis_task/views/business_requirement_gap_analysis_task.xml [hunk 10]
```diff
@@ -0,0 +1,54 @@
+<odoo>
+    <data>
```
Comment (elicoidal) [inline business_requirement_gap_analysis_task/views/business_requirement_gap_analysis_task.xml]: add copyright information in XML files too

File: business_requirement_gap_analysis_task/views/business_requirement_gap_analysis_task.xml [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+            </field>
+        </record>
+        <record model="ir.ui.view" id="view_on_change_of_master_project">
+            <field name="name">warning view</field>
+            <field name="model">business.requirement</field>
+            <field name="priority">20</field>
+            <field name="arch" type="xml">
+                <form string="Warning">
+                    <header>
+                        Waning:changing master project
```
Comment (elicoidal) [inline business_requirement_gap_analysis_task/views/business_requirement_gap_analysis_task.xml]: Waning:changing master project
=> Warning: changing master project

File: business_requirement_gap_analysis_task/views/res_config.xml [hunk 12]
```diff
@@ -0,0 +1,33 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+    <data>
```
Comment (elicoidal) [inline business_requirement_gap_analysis_task/views/res_config.xml]: data tag is not necessary in v10

**Review by elicoidal — CHANGES_REQUESTED** [review]
Thanks!
Mostly details.
No tests yet

**Comment by JoJoJoJoJoJoJo** [comment]
@elicoidal @seb-elico @lonelysun
changes done, review plz

**Review by elicoidal — CHANGES_REQUESTED** [review]
Almost there!
A good thing would to add demo data (one task link to one BR)

**Comment by elicoidal** [comment]
@sysadminmatmoz @cubells Would you like to review this one?

File: business_requirement_gap_analysis_task/__init__.py [hunk 13]
```diff
@@ -0,0 +1,4 @@
+# -*- coding: utf-8 -*-
+
+from . import wizard
+from . import models
```
Comment (cubells) [inline business_requirement_gap_analysis_task/__init__.py]: A blank line is needed after this

File: business_requirement_gap_analysis_task/__manifest__.py [hunk 14]
```diff
@@ -0,0 +1,23 @@
+# -*- coding: utf-8 -*-
+# © 2017 Elico Corp (https://www.elico-corp.com).
```
Comment (cubells) [inline business_requirement_gap_analysis_task/__manifest__.py]: You must change `©` by Copyright

File: business_requirement_gap_analysis_task/data/business_requirement_demo.xml [hunk 15]
```diff
@@ -0,0 +1,20 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<odoo>
+    <data noupdate="1">
```
Comment (cubells) [inline business_requirement_gap_analysis_task/data/business_requirement_demo.xml]: It must be `odoo noupdate="1"`

File: business_requirement_gap_analysis_task/data/gap_analysis_task_demo.xml [hunk 16]
```diff
@@ -0,0 +1,45 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo>
+    <data noupdate="1">
```
Comment (cubells) [inline business_requirement_gap_analysis_task/data/gap_analysis_task_demo.xml]: Idem

File: business_requirement_gap_analysis_task/models/business.py [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+    def master_project_change(self):
+        for rec in self:
+            if rec.gap_analysis_task:
+                return {
+                    'warning': {
+                        'title': 'Master Project Changed',
+                        'message': 'The master project has been changed,'
+                                   ' if you want to change the '
+                                   'master project of gap analysis task, '
+                                   'you have to change it manually'
```
Comment (cubells) [inline business_requirement_gap_analysis_task/models/business.py]: An `.` must end message

File: business_requirement_gap_analysis_task/wizard/create_gap_task.xml [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+        </field>
+    </record>
+    <record id="launch_wizard_gap" model="ir.actions.act_window">
+        <field name="name">Gap Analysis</field>
+        <field name="res_model">br.gap_analysis_task</field>
+        <field name="view_type">form</field>
+        <field name="view_mode">form</field>
+        <field name="target">new</field>
+    </record>
+</odoo>
```
Comment (cubells) [inline business_requirement_gap_analysis_task/wizard/create_gap_task.xml]: an end blank line is neeed

File: business_requirement_gap_analysis_task/wizard/create_gap_task.xml [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+        <field name="priority">20</field>
+        <field name="arch" type="xml">
+            <form>
+                <group>
+                    <field name="estimated_time"/>
+                </group>
+                <footer>
+                <button name="apply" string="OK" type="object"
+                        class="oe_highlight"/>
+                or
```
Comment (cubells) [inline business_requirement_gap_analysis_task/wizard/create_gap_task.xml]: `or` is not needed by usability

**Review by cubells — CHANGES_REQUESTED** [review]
Minor changes.

Ready to approve.

**Comment by elicoidal** [comment]
@cubells thanks for the review!

**Comment by JoJoJoJoJoJoJo** [comment]
@cubells 
Thanks for the review!
Changes are made, review the pr plz.
cc @elicoidal  @seb-elico  @lonelysun

**Review by elicoidal — APPROVED** [review]

File: business_requirement_gap_analysis_task/models/business.py [hunk 20]
```diff
@@ -0,0 +1,32 @@
...
+# Copyright 2017 Elico Corp (https://www.elico-corp.com).
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo import models, fields, api
+
+
+class BusinessRequirementGapAnalysis(models.Model):
+    _inherit = 'business.requirement'
+
+    gap_analysis_task = fields.Many2one(
```
Comment (lonelysun) [inline business_requirement_gap_analysis_task/models/business.py]: change field name to gap_analysis_task_id, The fields of the many2one type end with _id

File: business_requirement_gap_analysis_task/models/res_config.py [hunk 21]
```diff
@@ -0,0 +1,39 @@
...
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo import models, fields, api
+
+
+class BusinessRequirementConfigSettings(models.TransientModel):
+    _name = 'business.requirement.gap.task.config.setting'
+    _inherit = 'res.config.settings'
+
+    gap_task_category = fields.Many2one(
```
Comment (lonelysun) [inline business_requirement_gap_analysis_task/models/res_config.py]: change field name to gap_analysis_task_id, The fields of the many2one type end with _id

**Review by lonelysun — CHANGES_REQUESTED** [review]
Change the field name and don't forget the reference section

**Comment by elicoidal** [comment]
@JoJoJoJoJoJoJo Can you attend @lonelysun comments?
@cubells Can you please review the changes?

**Comment by JoJoJoJoJoJoJo** [comment]
Done @lonelysun
cc @elicoidal

File: business_requirement_gap_analysis_task/data/business_requirement_demo.xml [hunk 22]
```diff
@@ -0,0 +1,22 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<!-- Copyright 2017 Elico Corp (https://www.elico-corp.com). -->
+<!-- License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html). -->
+<odoo noupdate="1">
+    <data>
```
Comment (cubells) [inline business_requirement_gap_analysis_task/data/business_requirement_demo.xml]: Tag `data` is not needed here. Remove.

File: business_requirement_gap_analysis_task/data/gap_analysis_task_demo.xml [hunk 23]
```diff
@@ -0,0 +1,47 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<!-- Copyright 2017 Elico Corp (https://www.elico-corp.com). -->
+<!-- License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html). -->
+<odoo noupdate="1">
+    <data>
```
Comment (cubells) [inline business_requirement_gap_analysis_task/data/gap_analysis_task_demo.xml]: Tag `data` is not needed here. Remove.

File: business_requirement_gap_analysis_task/models/business.py [hunk 24]
```diff
@@ -0,0 +1,32 @@
...
+        string='Gap Analysis Task',
+        help='Use this task to input all the timesheets for the preparation of'
+             ' this BR (meetings, solution design, estimation, etc.)'
+    )
+
+    @api.multi
+    @api.onchange('project_id')
+    def master_project_change(self):
+        for rec in self:
+            if rec.gap_analysis_task_id:
```
Comment (cubells) [inline business_requirement_gap_analysis_task/models/business.py]: You can improve these lines with this one:
`if self.mapped('rec.gap_analysis_task_id'):`

**Review by cubells — CHANGES_REQUESTED** [review]
Minor changes.

File: business_requirement_gap_analysis_task/models/business.py (hunk 24 above)
Comment (cubells) [reply business_requirement_gap_analysis_task/models/business.py]: better:

`for rec in self.mapped('gap_analysis_task_id'):`

**Review by lonelysun — APPROVED** [review]

**Comment by elicoidal** [comment]
@cubells Please review again the corrections

**Review by cubells — APPROVED** [review]
LGTM

code review and test.

Goog job @JoJoJoJoJoJoJo

**Comment by elicoidal** [comment]
Thanks @cubells

**Pull request merged**